File: app/src/main/java/com/software/jgodort/marvelheroes/dependencyInjection/ApplicationComponent.java
package com.software.jgodort.marvelheroes.dependencyInjection;

import com.software.jgodort.marvelheroes.business.Interactors.impl.GetSuperHeroesInteractorImpl;
import com.software.jgodort.marvelheroes.network.MarvelSampleRepositoryImpl;
import com.software.jgodort.marvelheroes.presentation.presenter.impl.HeroesPresenterImpl;
import com.software.jgodort.marvelheroes.presentation.ui.HeroesActivity;
import com.software.jgodort.marvelheroes.presentation.ui.adapters.HeroesAdapter;

import javax.inject.Singleton;

import dagger.Component;

/**
 * Component to define the dependency injection to layers.
 * Created by javie on 07/08/2017.
 */

@Singleton
@Component(modules = {
        ApplicationModule.class,
        RetrofitModule.class,
        InteractorsModule.class,
        PresentersModule.class
})
public interface ApplicationComponent {

    void inject(HeroesAdapter adapter);

    void inject(HeroesActivity heroesActivity);

    void inject(HeroesPresenterImpl heroesPresenter);

    void inject(MarvelSampleRepositoryImpl marvelSampleRepository);

    void inject(GetSuperHeroesInteractorImpl getSuperHeroesInteractor);
}

File: app/src/main/java/com/software/jgodort/marvelheroes/dependencyInjection/component/ApplicationComponent.kt
package com.software.jgodort.marvelheroes.dependencyInjection.component

import android.app.Application
import com.software.jgodort.marvelheroes.MarvelSampleApp
import com.software.jgodort.marvelheroes.dependencyInjection.module.ActivityBindingModule
import com.software.jgodort.marvelheroes.dependencyInjection.module.ApplicationModule
import com.software.jgodort.marvelheroes.dependencyInjection.scopes.PerApplication
import dagger.BindsInstance
import dagger.Component
import dagger.android.support.AndroidSupportInjectionModule

@PerApplication
@Component(
    modules = [ActivityBindingModule::class,
      ApplicationModule::class,
      AndroidSupportInjectionModule::class]
)
interface ApplicationComponent {

  @Component.Builder
  interface Builder {
    @BindsInstance fun application(application: Application): Builder
    fun build(): ApplicationComponent
  }

  fun inject(app: MarvelSampleApp)
}